checkbox-provider-ce-oem/bin/digital_io_loopback_serial_test.py

```diff
@@ -0,0 +1,152 @@
+#!/bin/python3
+import serial
+import time
+import binascii
+from argparse import ArgumentParser
+
+
+def calculate_crc(*data):
+    crc_sum = 0x00
+    for byte in data:
+        crc_sum ^= byte
+        for _ in range(0, 8):
+            tmp_int = 7 if (crc_sum & 0x80) else 0
+            crc_sum = (crc_sum << 1) ^ tmp_int
+
+            if crc_sum > 255:
+                crc_sum -= 256
+    return crc_sum
+
+
+class DigitalIOSerialController():
+
+    TEST_STATES = [0, 1]
+    # DIGITAL_IN_PINS = [6, 7, 8, 9]
+    # DIGITAL_OUT_PINS = [2, 3, 4, 5]
+    PREFIX_BYTE = 68 # equal to 0x44
+
+    def __init__(self, port):
+        self.conn = serial.Serial("/dev/{}".format(port), timeout=2)
+
+    def generate_message(self, *data):
+        """Generate command with CRC checksum
+
+        Args:
+            *data (list): paritial bytes to communicate to serial console
+
+        Returns:
+            messages (bytearray): a command to communicate to serial console
+        """
+        crc_sum = calculate_crc(*data)
+        data = list(data)
+        data.append(crc_sum)
+
+        return bytearray(data)
+
+    def read_digital_in(self, pin_num):
+        """Read the value from digital input port
+        Will issue a command to serial console with following format
+        {function_byte} {pin_byte} {crc_sum}
+
+        Args:
+            pin_num (str): the port ot DI
+
+        Returns:
+            value (str): the value of DI
+        """
+        print("\n# Read value from register byte {}..".format(pin_num))
```

Comment:
I noticed that you wrote with `fstring` in other places and `format` in here, is there any reason for you to wrote it in two different way in these python scripts?

Reply:
fixed in latest commit